File: src/main/java/com/joecollins/bindings/BindingReceiver.java
package com.joecollins.bindings;

import java.util.function.Consumer;
import java.util.function.Function;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;

public class BindingReceiver<T> extends Bindable<BindingReceiver.Property> {
  enum Property {
    PROP
  }

  private T value;

  public BindingReceiver(Binding<? extends T> binding) {
    binding.bindLegacy(
        v -> {
          value = v;
          onPropertyRefreshed(Property.PROP);
        });
  }

  public Binding<T> getBinding() {
    return getBinding(Function.identity());
  }

  public <U> Binding<U> getBinding(Function<T, U> func) {
    return Binding.propertyBinding(this, t -> func.apply(t.value), Property.PROP);
  }

  public <U> Binding<U> getFlatBinding(Function<T, Binding<U>> func) {
    return new Binding<>() {
      private Binding<T> topBinding;
      private Binding<U> subBinding;
      private Consumer<U> consumer;

      @Override
      public U getValue() {
        return func.apply(value).getValue();
      }

      @Override
      public void bind(@NotNull Function1<? super U, Unit> onUpdate) {
        bindLegacy(onUpdate::invoke);
      }

      @Override
      public void bindLegacy(Consumer<U> onUpdate) {
        if (consumer != null) {
          throw new IllegalStateException("Binding already in use");
        }
        consumer = onUpdate;
        topBinding = getBinding();
        topBinding.bindLegacy(
            t -> {
              if (subBinding != null) {
                subBinding.unbind();
              }
              subBinding = func.apply(t);
              subBinding.bindLegacy(consumer);
            });
      }

      @Override
      public void unbind() {
        topBinding.unbind();
        if (subBinding != null) {
          subBinding.unbind();
        }
        consumer = null;
      }
    };
  }

  public T getValue() {
    return value;
  }
}

File: src/main/java/com/joecollins/bindings/BindingReceiver.kt
package com.joecollins.bindings

class BindingReceiver<T>(binding: Binding<T>) {

    private val _value = WrappedValue<T>()

    init {
        binding.bind { _value.value = it }
    }

    private enum class Property { VALUE }

    private class WrappedValue<T> : Bindable<WrappedValue<T>, Property>() {
        private var _value: T? = null

        var value: T
        get() {
            @Suppress("UNCHECKED_CAST")
            return _value as T
        }
        set(v) {
            _value = v
            onPropertyRefreshed(Property.VALUE)
        }
    }

    val value: T get() = _value.value

    fun getBinding(): Binding<T> = getBinding { it }

    fun <U> getBinding(func: (T) -> U): Binding<U> = Binding.propertyBinding(_value, { func(it.value) }, Property.VALUE)

    fun <U> getFlatBinding(func: (T) -> Binding<U>): Binding<U> {
        val me = this
        return object : Binding<U> {
            private var topBinding: Binding<T>? = null
            private var subBinding: Binding<U>? = null

            override fun bind(onUpdate: (U) -> Unit) {
                topBinding = me.getBinding()
                topBinding!!.bind {
                    subBinding?.unbind()
                    subBinding = func(it)
                    subBinding!!.bind(onUpdate)
                }
            }
            override fun unbind() {
                topBinding?.unbind()
                subBinding?.unbind()
            }
        }
    }
}
